autoconfigure/storage-elasticsearch-aws/src/main/java/zipkin/autoconfigure/storage/elasticsearch/aws/AWSSignatureVersion4.java

```diff
@@ -13,128 +13,222 @@
+  final byte[] regionBytes;
+  final HostsSupplier hostsSupplier;
   final String service;
+  final byte[] serviceBytes;
   final AWSCredentials.Provider credentials;
 
-  AWSSignatureVersion4(String region, String service, AWSCredentials.Provider credentials) {
+  AWSSignatureVersion4(Client<HttpRequest, HttpResponse> delegate, String region,
+      HostsSupplier hostsSupplier, AWSCredentials.Provider credentials) {
+    super(delegate);
     if (region == null) throw new NullPointerException("region == null");
-    if (service == null) throw new NullPointerException("service == null");
     if (credentials == null) throw new NullPointerException("credentials == null");
     this.region = region;
-    this.service = service;
+    this.regionBytes = region.getBytes(UTF_8);
+    this.hostsSupplier = hostsSupplier;
+    this.service = "es";
+    this.serviceBytes = service.getBytes(UTF_8);
     this.credentials = credentials;
   }
 
-  @Override
-  public Response intercept(Chain chain) throws IOException {
-    Request input = chain.request();
-    Request signed = sign(input);
-    Response response = chain.proceed(signed);
-    if (response.code() == 403) {
-      try (ResponseBody body = response.body()) {
-        JsonReader message = enterPath(JsonReader.of(body.source()), "message");
-        if (message != null) throw new IllegalStateException(message.nextString());
+  volatile String host;
+
+  String host() {
+    if (host == null) {
+      synchronized (this) {
+        if (host == null) {
+          // We currently cannot access the hostname during decoration. Instead, we proceed knowing
+          // that our implementation of AWS ES only returns a single endpoint.
+          List<String> hosts = hostsSupplier.get();
+          if (hosts.isEmpty()) throw new RuntimeException("No AWS ES hosts!");
+          if (hosts.size() > 1) throw new RuntimeException("Expected one AWS ES host: " + hosts);
+          host = URI.create(hosts.get(0)).getHost();
+        }
       }
-      throw new IllegalStateException(response.toString());
     }
-    return response;
+    return host;
+  }
+
+  @Override public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
+    return HttpResponse.from(
+        req.aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc())
+            .thenCompose(aggReg -> {
+              try {
+                AggregatedHttpRequest signed = sign(ctx, aggReg);
+                return delegate().execute(ctx, HttpRequest.of(signed)).aggregate();
+              } catch (Exception e) {
+                CompletableFuture<AggregatedHttpResponse> future = new CompletableFuture<>();
+                future.completeExceptionally(e);
+                return future;
+              }
+            })
+            .thenApply(aggResp -> {
+              if (!aggResp.status().equals(HttpStatus.FORBIDDEN)) {
+                return HttpResponse.of(aggResp);
+              }
+              String body = aggResp.contentUtf8();
```

Comment:
If we switch above to pooled, will need to `try {} finally { ReferenceCountUtil.safeRelease(aggResp.content())`

We could do our normal `InputStream` dance too but probably can wait until after `toInputStream` is fixed.